### PR TITLE
Preserve array for permitted subtype accross alloc

### PR DESCRIPTION
**Pull request by tajila**
Preserve array for permitted subtype accross alloc

Preserve the string array for permitted subtypes across array element allocations.

Related to: https://github.com/eclipse-openj9/openj9/issues/17254

### CODE REVIEW

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

**Comment by tajila** [comment]
jenkins test sanity.openjdk amac,xlinux,plinux,zlinux jdk

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

File: runtime/jcl/common/reflecthelp.c [hunk 1]
```diff
@@ -2022,33 +2022,42 @@ permittedSubclassesHelper(JNIEnv *env, jobject cls)
...
 		}
 
 		J9JAVAARRAYOFOBJECT_STORE(vmThread, stringArrayObject, index, nameString);
 	}
-	goto done;
 
-heapoutofmemory:
-	vmFuncs->setHeapOutOfMemoryError(vmThread);
-	goto done;
+	result = vmFuncs->j9jni_createLocalRef(env, (j9object_t)stringArrayObject);
```
Comment (JasonFengJ9) [inline runtime/jcl/common/reflecthelp.c:2051]:
```suggestion
	result = vmFuncs->j9jni_createLocalRef(env, stringArrayObject);
```
The cast is not required now.

**Comment by pshipton** [comment]
Oops, should have been:

jenkins test sanity.openjdk amac,xlinux,plinux,zlinux jdk17

**Comment by pshipton** [comment]
The running PR builds are:
https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/3719/
https://openj9-jenkins.osuosl.org/job/PullRequest-OpenJ9/3720/

**Comment by tajila** [comment]
Only failure was [ImplicitStringConcatShapes](https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.openjdk_s390x_linux_Personal/18/testReport/junit/java_lang_String_concat_ImplicitStringConcatShapes/java/ImplicitStringConcatShapes/)

**Comment by tajila** [comment]
@JasonFengJ9 Ive updated the PR

**Comment by tajila** [comment]
@babsingh Please review merge this PR

**Review by babsingh — APPROVED** [review]
[PR builds](https://github.com/eclipse-openj9/openj9/pull/17299#issuecomment-1526080870) look good. [ImplicitStringConcatShapes](https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.openjdk_s390x_linux_Personal/18/testReport/java_lang_String_concat_ImplicitStringConcatShapes/java/ImplicitStringConcatShapes/) is a known failure: https://github.com/eclipse-openj9/openj9/issues/17203.

**Pull request merged**